designs-for-feedback/trigger-workflows-in-pipeline.org

```diff
@@ -0,0 +1,69 @@
+* Request for Comment: Trigger an individual workflow
+** Problem statement
+Users want to be able to trigger a pipeline, selecting a specific workflow in
+the configuration to run. They also need a way for such manually triggered
+workflows to not always run when the pipeline is triggered without specifying a
+particular workflow.
+
+** Proposed solution
+We will add new semantics to workflows configuration which will use pipeline
+parameters (and potentially other types of values in the future) to determine
+whether a particular workflow should run within a pipeline. Standard pipeline
+parameters can then be passed in to select which workflows should be run.
+
+** Design options for configuration of workflows
+If a pipeline configuration contains multiple workflows the behavior today is to
+run them both immediately upon triggering. With the ability to trigger a
+particular workflow via the API, users will likely want to be able to configure
+workflows intended to be run only when explicitly triggered and not run when a
+normal webhook from GitHub or other VCS system is received upon a push to the
+repo. Thus, we need to introduce some mechanism to declare under what
+circumstances certain workflows in a configuration should be run.
+
+The currently planned solution uses the existing pipeline parameters feature
+coupled with some new semantics within workflows to decide whether a workflow
+will be run.
+
+#+BEGIN_SRC yaml
+  parameters:
+    run_integration_tests:
+      type: boolean
+      default: false
+
+  workflows:
+    integration_tests:
+      when:
+        condition: << pipeline.parameters.run_integration_tests >>
+      jobs:
+        - tests
```

Comment:
Can we see allowing `when` on an individual job as well?

Reply:
As long as we are very clear about what happens to the workflow this could be a powerful config tool.
A job that won't be run due to a `when`/`unless` condition will be removed from the workflow graph which will also remove any descendant of that job. It will be possible to accidentally remove large chunks of a workflow graph and it may not be obvious to a customer why that happened.

I'd also like us to think really hard about how to incorporate branch filters into the `when`/`unless`, having both complicates config, and these conditions give us an opportunity to tidy up and consolidate.

Reply:
With the current UI I can also see the problem that you can't see the config in the workflow view and there is no pipeline view, so if you accidentally remove the root job of a workflow the whole thing is gone without us being able to surface that information.

I think branch filters fit very neatly into `when`/`unless`:
```yaml:
when:
  branch: master
...
unless:
  branch: canary/*
```
Question is if we want to allow combinations of several "condition types", like both `condition` and `branch` in the same `when`. I can see the value, but I'm not sure how complex that would end up being.

Reply:
Right, that's exactly what we need to consider. I can see value in being able to say "run this job when doing integration tests, but only on `master`", but that raises the question of how multiple "condition types" combine.

With a `when` the "obvious" answer is to AND them, and a similar bit of thinking leads to ORing in an `unless` but since those are both implied it makes things harder for the person writing the config. An explicit expression system would be nice but feels out of scope.